### PR TITLE
feat: support use storybook to dev rslib module federation format assets and also support hmr

### PR DESCRIPTION
**Pull request by nyqykk**
## Summary
Support use storybook to dev rslib module federation format assets and also support hmr
Many capabilities are integrated here.
https://github.com/rspack-contrib/storybook-rsbuild/pull/165
https://github.com/module-federation/core/pull/3131

## Related Links
- [ ] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/cli/commands.ts
Comment (fi3ework) [reply packages/core/src/cli/commands.ts:88]: What is the difference between `dev` and `build --watch`. This is MF specified, it's not likely to merge into the common command, consider moving it to the MF plugin?

File: examples/module-federation/mf-react-component/tsconfig.json
Comment (fi3ework) [inline examples/module-federation/mf-react-component/tsconfig.json:11]: It's better to not put stories inside `src` folder, but beside it.

File: examples/module-federation/mf-react-component/tsconfig.json
Comment (fi3ework) [inline examples/module-federation/mf-react-component/tsconfig.json:8]: What it does for?

File: packages/core/src/cli/commands.ts
Comment (nyqykk) [reply packages/core/src/cli/commands.ts:88]: build --watch is also production mode and hmr runtime will not be injected into browser.
use dev command can make hmr work normally

File: examples/module-federation/mf-react-component/tsconfig.json
Comment (nyqykk) [reply examples/module-federation/mf-react-component/tsconfig.json:8]: ensure type check in `*.stories.ts`

File: packages/core/src/cli/commands.ts
Comment (nyqykk) [reply packages/core/src/cli/commands.ts:88]: `dev` command only used in rslib, so this may be add into rslib repo

File: examples/module-federation/mf-react-component/tsconfig.json
Comment (nyqykk) [reply examples/module-federation/mf-react-component/tsconfig.json:11]: i just follow this repo, do we need to put it beside `src` folder? https://github.com/rspack-contrib/storybook-rsbuild/tree/main/sandboxes/react-16

Comment (fi3ework) [reply examples/module-federation/mf-react-component/tsconfig.json:11]: It's better, will reduce the code to ship to NPM. The code in that repository is forked from Storybook's official sandbox repository, while they're not targeting library but app.

File: examples/module-federation/mf-react-component/tsconfig.json
Comment (fi3ework) [reply examples/module-federation/mf-react-component/tsconfig.json:8]: Why does it help?

Comment (nyqykk) [reply examples/module-federation/mf-react-component/tsconfig.json:8]: Storybook app like a host app in mf, it will consume rslib module mf assets.  And a remote app will build its own types for host app to ensure type check in host app. If we do not add this in `tsconfig` this import can not pass type check because this is a remote module [code](https://github.com/web-infra-dev/rslib/pull/349/files#diff-d00fa483b327ce5e4ceaac742ff6a53dfa43eeef21cca15d6a7170b848c3d819R2)

File: examples/module-federation/mf-react-component/package.json
Comment (fi3ework) [inline examples/module-federation/mf-react-component/package.json:17]: what does server do here?

File: packages/core/src/cli/commands.ts
Comment (fi3ework) [reply packages/core/src/cli/commands.ts:88]: > dev command only used in rslib, so this may be add into rslib repo

How to use with the `dev` command? Can it work without Storybook?

File: examples/module-federation/mf-react-component/package.json
Comment (nyqykk) [reply examples/module-federation/mf-react-component/package.json:17]: `serve` command is to test assets of mf format is correct, e2e test will use this command https://github.com/web-infra-dev/rslib/blob/main/tests/playwright.config.ts#L17

File: packages/core/src/cli/commands.ts
Comment (nyqykk) [reply packages/core/src/cli/commands.ts:88]: i will write it into doc if we make sure this command can be added into rslib repo, now we use `rslib dev:mf` can start the command. We will not use this command without storybook in normal

Comment (fi3ework) [reply packages/core/src/cli/commands.ts:91]: > We will not use this command without storybook in normal

1. If it's Storybook only, we could add it to https://github.com/rspack-contrib/storybook-rsbuild/blob/main/packages/addon-rslib/package.json. I think we only need to export `composeCreateRsbuildConfig` from Rslib core and finish all other things in the addon.
2. If not, such as use with Rspress ( cc @Timeless0911 ), we could keep `initDevMFRsbuild` inside Rslib core, the and the rest part to Rspress/Storybok. However, since `initDevMFRsbuild` is quite simple, I'm also fine Rspress/Storybook duplicate it themself.

Comment (nyqykk) [reply packages/core/src/cli/commands.ts:91]: yeah, i will follow the sugguestion 1

Comment (Timeless0911) [reply packages/core/src/cli/commands.ts:91]: > `build --watch is also production mode`

Rslib use fixed `mode: production` yet. I wonder how hmr runtime injected into browser, by NODE_ENV?

Comment (nyqykk) [reply packages/core/src/cli/commands.ts:91]: i will change mode to `development` and use `rsbuild.startDevServer ` to start dev mode to ensure hmr https://github.com/web-infra-dev/rslib/pull/349/files#diff-4114b8f08217be27efd35a3891857cfabc6eb75aee8ab39ba304e4d1b1095323R1251

Comment (nyqykk) [reply packages/core/src/cli/commands.ts:91]: i'm trying to move this to rsbuild storybook addon

Comment (fi3ework) [reply packages/core/src/cli/commands.ts:91]: Just to ensure: IMO, user need to use @module-federation/storybook-addon alongside storybook-addon-rslib?

Comment (fi3ework) [reply packages/core/src/cli/commands.ts:91]: What can user do only with `@module-federation/storybook-addon`?

Comment (nyqykk) [reply packages/core/src/cli/commands.ts:91]: app project can only use `@module-federation/storybook-addon`, lib can not

Comment (fi3ework) [reply packages/core/src/cli/commands.ts:91]: It's a little confused for me now. We should demonstrates the relationships in the document.

**Comment by nyqykk** [comment]
Many capabilities are integrated here.
https://github.com/rspack-contrib/storybook-rsbuild/pull/165
https://github.com/module-federation/core/pull/3131

**Comment by fi3ework** [comment]
I'll release a new version of storybook-addon-rsllib.

File: examples/module-federation/mf-react-component/README.md
Comment (fi3ework) [inline examples/module-federation/mf-react-component/README.md:9]:
```suggestion
1. Start remote module which is loaded by Rslib module.
```

File: examples/module-federation/mf-react-component/README.md
Comment (fi3ework) [inline examples/module-federation/mf-react-component/README.md:15]:
```suggestion
2. Use Storybook to development component.
```

File: examples/module-federation/mf-react-component/package.json
Comment (fi3ework) [inline examples/module-federation/mf-react-component/package.json:8]: `"types"` must be moved to the first field under `"."`.

File: website/docs/en/guide/advanced/module-federation.mdx
Comment (fi3ework) [inline website/docs/en/guide/advanced/module-federation.mdx:172]:
```suggestion
#### 5. Start Storybook app
```

File: website/docs/en/guide/advanced/module-federation.mdx
Comment (fi3ework) [inline website/docs/en/guide/advanced/module-federation.mdx:174]:
```suggestion
There you go, start Storybook with `npx storybook dev`.
```

Comment (fi3ework) [reply website/docs/en/guide/advanced/module-federation.mdx:174]: I've also fixed the wrong command where you copied it.

**Comment by fi3ework** [comment]
The lockfile changed too much, you could try to reset lockfile to main HEAD first then re-run `pnpm i` to update it.

File: examples/module-federation/mf-react-component/stories/index.stories.tsx
Comment (nyqykk) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: this is not an expect error, in local we have @mf-types for type check. So if we change this to expect error, we will receive error hint to remove this comment, this can not be changed

**Comment by nyqykk** [comment]
> The lockfile changed too much, you could try to reset lockfile to `main` branch first then re-run `pnpm i` to update it.

it seems normal change, you can also have a try, i get the same result

**Comment by fi3ework** [comment]
The initial investigation appears to be caused by peer dep doppelganger.

**Comment by fi3ework** [comment]
@Timeless0911 PTAL https://github.com/web-infra-dev/rslib/pull/349/commits/0aa8efc94790a258b5357ab844a837827ba68cf4. I disabled `auto-install-peers`.

File: examples/module-federation/mf-react-component/stories/index.stories.tsx
Comment (fi3ework) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: So is it an TS error? If not, why there's an `@ts-ignore`?

**Comment by Timeless0911** [comment]
> @Timeless0911 PTAL [0aa8efc](https://github.com/web-infra-dev/rslib/commit/0aa8efc94790a258b5357ab844a837827ba68cf4). I disabled `auto-install-peers`.

LGTM, it's more strict now. And is this PR going to merge right now?

File: examples/module-federation/mf-react-component/stories/index.stories.tsx
Comment (nyqykk) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: yes, you can consider this is a ts error in ci env because we set @mf-types into .gitignore, when storybook build, type check will throw error

Comment (fi3ework) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: How to generate in CI env just like local.

Comment (nyqykk) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: remove @mf-types from `.gitignore`

Comment (nyqykk) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: like this folder https://github.com/module-federation/core/pull/3131/files#diff-cf94c81add15f5e791dc7eb01c930f05f645750b1f6fda34ce6727b463f22fb0

Comment (nyqykk) [reply examples/module-federation/mf-react-component/stories/index.stories.tsx:1]: files in this folder are auto generated by mf

File: packages/core/src/cli/commands.ts
Comment (fi3ework) [inline packages/core/src/cli/commands.ts:115]:
```suggestion
    .description(`start Rsbuild dev server of Module Federation format`)
```

File: packages/core/src/mf.ts
Comment (fi3ework) [inline packages/core/src/mf.ts:7]: IIUC, after this is merged, we could remove https://github.com/rspack-contrib/storybook-rsbuild/blob/d574bc218836299aa07ca60b6bb3a90c6a01ec27/packages/addon-rslib/src/mf.ts logic?

File: packages/core/src/cli/commands.ts
Comment (Timeless0911) [inline packages/core/src/cli/commands.ts:114]: add this command to docs https://lib.rsbuild.dev/guide/basic/cli

File: packages/core/src/mf.ts
Comment (nyqykk) [reply packages/core/src/mf.ts:7]: yes, i have done this, and will open a pr

**Comment by fi3ework** [comment]
I will add `--lib` args afterwards.

**Review by fi3ework — APPROVED** [review]

**Pull request merged**
